### PR TITLE
Handle json encoding of V1Pod in task callback

### LINKED ISSUE #27593: Object of type V1Pod is not JSON serializable after detecting zombie jobs cause Scheduler CrashLoopBack
### Apache Airflow version

2.4.2

### What happened

Some dags have tasks with pod_override in executor_config become zombie tasks. Airflow Scheduler run and crash with exception:

```
[2022-11-10T15:29:59.886+0000] {scheduler_job.py:1526} ERROR - Detected zombie job: {'full_filepath': '/opt/airflow/dags/path.py', 'processor_subdir': '/opt/airflow/dags', 'msg': "{'DAG Id': 'dag_id', 'Task Id': 'taskid', 'Run Id': 'manual__2022-11-10T10:21:25.330307+00:00', 'Hostname': 'hostname'}", 'simple_task_instance': <airflow.models.taskinstance.SimpleTaskInstance object at 0x7fde9c91dcd0>, 'is_failure_callback': True}
[2022-11-10T15:29:59.887+0000] {scheduler_job.py:763} ERROR - Exception when executing SchedulerJob._run_scheduler_loop
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/jobs/scheduler_job.py", line 746, in _execute
    self._run_scheduler_loop()
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/jobs/scheduler_job.py", line 878, in _run_scheduler_loop
    next_event = timers.run(blocking=False)
  File "/usr/local/lib/python3.7/sched.py", line 151, in run
    action(*argument, **kwargs)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/utils/event_scheduler.py", line 37, in repeat
    action(*args, **kwargs)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/utils/session.py", line 75, in wrapper
    return func(*args, session=session, **kwargs)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/jobs/scheduler_job.py", line 1527, in _find_zombies
    self.executor.send_callback(request)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/executors/base_executor.py", line 400, in send_callback
    self.callback_sink.send(request)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/utils/session.py", line 75, in wrapper
    return func(*args, session=session, **kwargs)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/callbacks/database_callback_sink.py", line 34, in send
    db_callback = DbCallbackRequest(callback=callback, priority_weight=10)
  File "<string>", line 4, in __init__
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/orm/state.py", line 480, in _initialize_instance
    manager.dispatch.init_failure(self, args, kwargs)
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/util/langhelpers.py", line 72, in __exit__
    with_traceback=exc_tb,
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 207, in raise_
    raise exception
  File "/home/airflow/.local/lib/python3.7/site-packages/sqlalchemy/orm/state.py", line 477, in _initialize_instance
    return manager.original_init(*mixed[1:], **kwargs)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/models/db_callback_request.py", line 46, in __init__
    self.callback_data = callback.to_json()
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/callbacks/callback_requests.py", line 89, in to_json
    return json.dumps(dict_obj)
  File "/usr/local/lib/python3.7/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
  File "/usr/local/lib/python3.7/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/local/lib/python3.7/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/usr/local/lib/python3.7/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type V1Pod is not JSON serializable
``` 

### What you think should happen instead

DbCallbackRequest should do to_json successfully

### How to reproduce

Start airflow with KubernetesExecutor
Make zombie task.

### Operating System

docker.io/apache/airflow:2.4.2

### Versions of Apache Airflow Providers

_No response_

### Deployment

Official Apache Airflow Helm Chart

### Deployment details

_No response_

### Anything else

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by dstandish**
Resolves #27593

### CODE REVIEW

File: airflow/models/taskinstance.py [hunk 1]
```diff
@@ -2572,12 +2572,13 @@ def __eq__(self, other):
...
-            if key in ["start_date", "end_date"]:
-                val = new_dict[key]
-                if not val or isinstance(val, str):
-                    continue
-                new_dict.update({key: val.isoformat()})
+        for key in ["start_date", "end_date"]:
+            val = new_dict[key]
+            if not val or isinstance(val, str):
+                continue
+            new_dict.update({key: val.isoformat()})
```
Comment (uranusjr) [inline airflow/models/taskinstance.py]:
```suggestion
            new_dict[key] = val.isoformat()
```

**Review by uranusjr — APPROVED** [review]

**Comment by dstandish** [comment]
chopped one more line here @uranusjr  if you want to doublecheck https://github.com/apache/airflow/pull/27609/commits/35839ea0db3a7f7c900adcaa156dc42e2f4f69c3

File: airflow/models/taskinstance.py [hunk 2]
```diff
@@ -2573,10 +2573,9 @@ def __eq__(self, other):
     def as_dict(self):
         new_dict = dict(self.__dict__)
         for key in ["start_date", "end_date"]:
-            val = new_dict[key]
-            if not val or isinstance(val, str):
-                continue
-            new_dict[key] = val.isoformat()
+            val = new_dict.get(key)
+            if hasattr(val, "isoformat"):
```
Comment (uranusjr) [inline airflow/models/taskinstance.py]: Would it be better to do a `isinstance(val, datetime.date)`?

Comment (dstandish) [reply airflow/models/taskinstance.py]: i would not trust that with all our datetime shenanigans... would you ? :)

Comment (uranusjr) [reply airflow/models/taskinstance.py]: 🤷

Comment (dstandish) [reply airflow/models/taskinstance.py]: We'll, maybe all of our date times inherit from that? If so it'd be fine. But if not we might miss stringifying it is the only thing

Comment (uranusjr) [reply airflow/models/taskinstance.py]: datetime does inherit from date; pendulum DateTime and Date inherits from datetime and date respectively.

Comment (dstandish) [reply airflow/models/taskinstance.py]: Ok then, can do

**Comment by akizminet** [comment]
Why don't we use `AirflowJsonEncoder` for `to_json` method like this PR: https://github.com/apache/airflow/issues/27593 ? It's more cleaner

**Comment by dstandish** [comment]
> Why don't we use `AirflowJsonEncoder` for `to_json` method like this PR: #11952 ? It's more cleaner

Seemed like a good idea so I explored it just now.  The problem is that AirflowJsonEncoder is only one way -- there's no AirflowJsonDecoder that will get you back the same object.  For example, when encoding a pod, it doesn't store anywhere the information that the json came from a v1pod so when decoding the json you would just get back dict of plain json-compatible objects.   So, we cannot use this approach. 

HOWEVER... with some adjustments, we can use BaseSerialization.  Not sure if this is any better but, i've taken a crack at it.  

@uranusjr what do you think is better?  Currently PR is converting to use BaseSerialization.  The original approach remains visible at https://github.com/apache/airflow/compare/main..15987961421ea5aa3c05b429c068cec85c55b774

File: airflow/models/taskinstance.py [hunk 3]
```diff
@@ -2571,6 +2571,7 @@ def __eq__(self, other):
         return NotImplemented
 
     def as_dict(self):
+        warnings.warn("This method is deprecated.  Use BaseSerialization.serialize.", DeprecationWarning)
```
Comment (uranusjr) [inline airflow/models/taskinstance.py]: Probably should have `stacklevel=2` to make the warning show the previous stack?

**Review by potiuk — APPROVED** [review]

**Pull request merged**